### ISSUE TITLE
Argument parser is a mess

### ISSUE THREAD

**Issue by loicgelle** [SRI-CSL/gllvm#4]
Try to simulate something like that and print the output from the parser:
```
gclang++ -pthread -m64 -m64  -o /build/nodejs-6.11/out/Release/mksnapshot -Wl,--start-group /build/nodejs-6.11/out/Release/obj.target/mksnapshot/deps/v8/src/snapshot/mksnapshot.o /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_base.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_nosnapshot.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_libplatform.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicui18n.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_libbase.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicuucx.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicudata.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicustubdata.a -Wl,--end-group -static -ldl -lr
```

The parser gives you that:
```
{[-pthread -m64 -m64 -o /build/nodejs-6.11/out/Release/mksnapshot -Wl,--start-group /build/nodejs-6.11/out/Release/obj.target/mksnapshot/deps/v8/src/snapshot/mksnapshot.o /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_base.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_nosnapshot.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_libplatform.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicui18n.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_libbase.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicuucx.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicudata.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicustubdata.a -Wl,--end-group -static -ldl -lrt] [] [/build/nodejs-6.11/out/Release/obj.target/mksnapshot/deps/v8/src/snapshot/mksnapshot.o /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_base.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_nosnapshot.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_libplatform.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicui18n.a /build/nodejs-6.11/out/Release/obj.target/deps/v8/tools/gyp/libv8_libbase.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicuucx.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicudata.a /build/nodejs-6.11/out/Release/obj.target/tools/icu/libicustubdata.a] /build/nodejs-6.11/out/Release/mksnapshot [-pthread -m64 -m64 -Wl,--end-group] [-Wl,--start-group -static -ldl -lrt] false false false false false false false}
```

Two things here:
- For some reason `-Wl,--start-group` is considered as a compile-time flag whereas `-Wl,--end-group` is considered as a link-time flag
- In any case everything that is between these flags should be considered as both compile-time + link-time flags, or all the builds will fail.

@ianamason

**Comment by ianamason** [comment]
Looks fun... How many more gotchas like this are there? That's why I want to pay more attention to the clang source rather than just make things up as we go along.

**Comment by loicgelle** [comment]
@ianamason Understood why it fails sometimes but not always: golang iterates map in a random order. So when we write:

```
var argPatterns = map[string]flagInfo{
		`^.+\.(c|cc|cpp|C|cxx|i|s|S|bc)$`:       {0, pr.inputFileCallback},
		`^.+\.([fF](|[0-9][0-9]|or|OR|pp|PP))$`: {0, pr.inputFileCallback},
		`^.+\.(o|lo|So|so|po|a|dylib)$`:         {0, pr.objectFileCallback},
		`^.+\.dylib(\.\d)+$`:                    {0, pr.objectFileCallback},
		`^.+\.(So|so)(\.\d)+$`:                  {0, pr.objectFileCallback},
		`^-(l|L).+$`:                            {0, pr.linkUnaryCallback},
		`^-I.+$`:                                {0, pr.compileUnaryCallback},
		`^-D.+$`:                                {0, pr.compileUnaryCallback},
		`^-U.+$`:                                {0, pr.compileUnaryCallback},
		`^-Wl,.+$`:                              {0, pr.linkUnaryCallback},
		`^-W.*$`:                                {0, pr.compileUnaryCallback},
		`^-f.+$`:                                {0, pr.compileUnaryCallback},
		`^-rtlib=.+$`:                           {0, pr.linkUnaryCallback},
		`^-std=.+$`:                             {0, pr.compileUnaryCallback},
		`^-stdlib=.+$`:                          {0, pr.compileLinkUnaryCallback},
		`^-mtune=.+$`:                           {0, pr.compileUnaryCallback},
		`^--sysroot=.+$`:                        {0, pr.compileUnaryCallback},
		`^-print-prog-name=.*$`:                 {0, pr.compileUnaryCallback},
		`^-print-file-name=.*$`:                 {0, pr.compileUnaryCallback},
	}
```

we expect the case `^-W.*$` to be a fallback case after `^-Wl,.+$`, but it is not always true. Have to fix that.

**Comment by ianamason** [comment]
Good.

**Comment by loicgelle** [comment]
Fixed by 5e80caa + 2f2b64b altough I left a FIXME in the code. Handling object files seems to be a pain.
